rust/connlib/tunnel/src/sockets.rs

```diff
@@ -68,35 +66,40 @@ impl Sockets {
         self.socket_v6.as_ref().map(|s| s.socket.as_raw_fd())
     }
 
-    pub fn poll_send_ready(&self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
-        if let Some(socket) = self.socket_v4.as_ref() {
-            ready!(socket.poll_send_ready(cx))?;
+    /// Flushes all buffered data on the sockets.
+    ///
+    /// Returns `Ready` if the socket is able to accept more data.
+    pub fn poll_flush(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
```

Comment:
> Could this result in e.g. IPv6 traffic waiting for the IPv4 socket to flush even if the IPv6 socket is ready?

Technically yes. Both of them have to be ready because prior to reading a packet from the device, I don't know whether I'll need to send it via IPv4 or IPv6.

Reply:
> Do we have any perf tests that cover simultaneous tx+rx on IPv4+IPv6?

No. We would need two gateways, one connected via IPv4 and one via IPv6 and IPv6 doesn't work properly in docker so we don't have any tests for that.